### ISSUE TITLE
Cursor Color Contrast Logic is Too Sensitive

### ISSUE THREAD

**Issue by Neightro** [alacritty/alacritty#4564]
I have a color specified for my cursor. When hovering over text that has the 'reverse' style, the cursor switches to be the inverse of the color underneath it. My understanding, as informed by @kchibisov in #2398, is that this is caused by a feature intended to keep the cursor itself from blending in with text that is very similar in color. However, this functionality kicks in in some situations where it likely shouldn't.

## My alacritty.yml

[alacritty.yml.txt](https://github.com/alacritty/alacritty/files/5683408/alacritty.yml.txt)

## To Reproduce

I've found two ways of reproducing this. The first is, while running my config and any program that doesn't change the background color, use the mouse to select over the cursor, which should turn black despite not being similar in color to the highlighted text.

![image](https://user-images.githubusercontent.com/23202923/101971520-b7607180-3be6-11eb-98e3-ea611716e2f2.png)

Alternatively, should that approach not work for some reason:

1) Run `vim -u NONE test.c`
2) Type the text `int test`
3) Within Vim, run the command `syn on | hi Type cterm=reverse gui=reverse`
4) Hover over the keyword `int`, which should be highlighted in lime green. The cursor will turn black, and the character underneath green. (Image is not very accurate, but I can't stop the window from unfocusing when taking a screenshot.)

![image](https://user-images.githubusercontent.com/23202923/101991110-5c1e9580-3c5f-11eb-8893-c3801902aedd.png)

## Current Behavior

In many cases, the cursor changes color when hovering over text with the reverse style even if the color of the text is not similar to that of the cursor. My cursor is orange (#e19b14), though it will turn black over text that is green or teal, for instance.

This behavior has the side effect of making the cursor hard to see in many cases, such as when hovering over matching delimiters in Vim with 'matchparen'. This doesn't affect every color scheme, though; only ones using the 'reverse' style with a bright enough color.

## Expected Behavior

The cursor only changes color when the background is very similar in color to the cursor itself, such as white on very light gray. Ideally this behavior would never happen with Vim's 'matchparen' highlighting or other one-character highlights, though that's likely a different issue.

## System 

Manjaro and Arch Linux
alacritty 0.6.0 (04cbf767)
X11 (xorg-server 1.20.10-0)

Please let me know if you need any more information, and I will be happy to oblige.

**Comment by chrisduerr** [comment]
I mean #E19B14 will have extremely bad contrast with almost everything else out there, so I'm not sure what to do here since it seems to be doing its job. At least the contrast of that green to your cursor's yellow is extremely low.

What exactly do you want to change, just to lower the threshold? The current threshold is a contrast of `1.5`, with your yellow on green being `1.3`. The lowest possible value is `1.0`, the highest is `21`, while I think the lowest acceptable contrast for text according to the W3C is 3 with large text.

So I have very little doubt a contrast of 1.3 justifies switching out the colors. Even if we were to reduce it even more we'd have to use at most `1.2` for your example to work. That would accept `#ff0000` and `#dd0000` as having enough contrast, which I don't think I agree with.

@kchibisov what do you think?

**Comment by Neightro** [comment]
Out of interest, what algorithm does Alacritty use to calculate contrast, and how much does it take hue into consideration? I agree that `#ff0000` and `#dd0000` should always invert the cursor, but I can't help wondering, should for example `#e19b14` (orange) and `#56b6c2` (teal), which are near complements, realistically do the same? It's possible that I'm overestimating the actual visibility between these two colors; just food for thought.

Truthfully, the reason I created this issue is that it's been noticed the cursor is sometimes hard to see when Vim highlights matching parens (#2398), and @kchibisov and I narrowed down the root cause to this contrast logic. Perhaps there's a better way to approach the problem than this.

**Comment by chrisduerr** [comment]
Alacritty uses the W3C standard for color contrast: https://www.w3.org/TR/WCAG20/#contrast-ratiodef

**Comment by chrisduerr** [comment]
> Truthfully, the reason I created this issue is that it's been noticed the cursor is sometimes hard to see when Vim highlights matching parens (#2398)

If the contrast is too low, Alacritty will use inversion though. If that doesn't provide sufficient contrast, there's something wrong with your colors.

**Comment by AckslD** [comment]
@Neightro @chrisduerr I experimented a bit since I think I'm having the same issue (see also my comment [here](https://github.com/alacritty/alacritty/issues/2398#issuecomment-872059754)). It seems that whenever you set both `text` and `cursor` in `colors: cursor` then the cursor becomes invisible within search matches in vim. It doesn't matter what color you set them to. I try many different, having them different, the same etc. If you only set one of them then the cursor will be visible in some way but not the way, I want to fit the colorscheme.

**Comment by bagnaram** [comment]
I am able to replicate this issue with the following snippet:

```
on=$(tput smso)"no color, standout on"`fb=3;r=255;g=85;b=85;printf '\e[0;%s8;2;%s;%s;%sm colored, standout on ' "$fb" "$r" "$g" "$b"`$(tput rmso)$(tput sgr0)
off="no color, standout off"`fb=3;r=255;g=85;b=85;printf '\e[0;%s8;2;%s;%s;%sm colored, standout off ' "$fb" "$r" "$g" "$b"`$(tput sgr0)
echo $off"\n"$on
```

It appears to be a problem when using standout under certain color combinations. This especially becomes an issue when using cursors with style `standout`, they disappear in certain situations.
